src/transformers/models/patchtst/modeling_patchtst.py

```diff
@@ -0,0 +1,1958 @@
+    mask[:, :, :len_keep] = 0
+
+    # sort noise for each sample
+    ids_shuffle = torch.argsort(noise, dim=-1)  # ascend: small is keep, large is remove
+    ids_restore = torch.argsort(ids_shuffle, dim=-1)  # ids_restore: [bs x nvars x L]
+
+    mask = torch.gather(mask, dim=-1, index=ids_restore)
+    mask = mask.unsqueeze(-1).repeat(1, 1, 1, feat)  # mask: [bs x nvars x num_patches x patch_length]
+    if unmasked_channel_indices is not None:
+        mask[:, unmasked_channel_indices, :, :] = 0
+
+    xb_mask = xb.masked_fill(mask.bool(), mask_value)
+    return xb_mask, mask[..., 0]
+
+
+def forecast_masking(
+    xb: torch.Tensor,
+    patch_lengths: list,
+    mix_ratio: list = None,
+    unmasked_channel_indices: list = None,
+    mask_value: int = 0,
+    seed_number: Optional[int] = None,
+):
+    """forecast_masking Mask last K patches where K is from the patch_lengths list.
+    For every batch, distribute the patch lengths based on mix_ratio Ignore masks for column indices mentioned in
+    cv_channel_indices
+
+    Args:
+        xb (Tensor):
+            Input to mask [ bs x nvars x num_patch x patch_len] or [ bs x tsg1 x tag2 x nvars x num_patch x patch_len]
+        patch_lengths (list): List of patch lengths to mask in the end of the data.
+        mix_ratio (list, optional): List of weights to use for each patch length. For Ex.
+            if patch_lengths is [5,4] and mix_ratio is [1,1], then equal weights to both patch lengths. Defaults to
+            None.
+        unmasked_channel_indices (list, optional):
+            Control Variable channel indices. These channels will not be masked. Defaults to None.
+        mask_value (int, optional): Value to use for masking. Defaults to 0.
+        seed_number (int, optional): Value to set for the random seed.
+
+    Returns:
+        Tensor: xb_mask, masked input, same shape as input Tensor: Mask tensor of shape [bs x c x n] or [bs x tsg1 x
+        tsg2 x c x n]
+    """
+    if seed_number:
+        set_seed(seed_number)
+
+    if mix_ratio is None:
+        mix_ratio = [1 for t in patch_lengths]
+
+    batch_size, nvars, seq_len, feat = xb.shape
+    mask = torch.zeros(batch_size, nvars, seq_len, device=xb.device)
+
+    t_list = []
+    total_length = 0
+    total_ratio = sum(mix_ratio)
+
+    for i, j in zip(patch_lengths, mix_ratio):
+        if i <= 0 or i >= seq_len:
+            raise Exception("masked_patch_len should be greater than 0 and less than total patches.")
+        temp_len = int(batch_size * j / total_ratio)
+        t_list.append([i, j, temp_len])
+        total_length += temp_len
+
+    t_list = sorted(t_list, key=lambda x: x[2])
+
+    if total_length < batch_size:
+        t_list[0][2] = t_list[0][2] + (batch_size - total_length)
+    elif total_length > batch_size:
+        t_list[-1][2] = t_list[-1][2] + (total_length - batch_size)
+
+    b1 = 0
+    for p, _, l in t_list:
```

Comment:
unclear what p and l are here

Reply:
same, would be nice to use clearer names than p, b1 and b2